_posts/2018-02-05-unrealiable-network.md

```diff
@@ -0,0 +1,111 @@
+
+# When the Worst Happens: Tell You are Offline
+
+This tip applies especially for single page and mobile applications, in which data is requested from the server between page loads. If the requests start failing, your application is probably going to look like a bunch of loading bars or icons. Your user can switch between application pages or views, but none of them is loading because of lost network connection. This is not a good user experience. Even if we cannot prevent network requests from failing from time to time, we can at least inform the user about the situation. A simple modal dialog or information bar at the top of your application will do fine, as long as you remember to remove it when the connection has been re-established. If possible, you might also want to consider allowing the user to keep using the application offline and sending the changes to the server when possible.
+
+# Testing Apps By Simulating Unrealiable Network
+
+Experience has shown me that we should always test how our applications work in unrealiable or slow network. When developing and testing an application locally with locally installed server and database, there is almost no network latency at all and the connections are very quick. Thus, you do not really get the real end user experience.
+
+The possible problems that you do not see with fast connections can vary. A typical problem is for example a button, which when clicked, sends a network request, and is rendered as enabled for the whole time. In most cases, the user should not be able click the button when the network request is being processed, so the button should be rendered as disabled. When the request is processed very quickly on local development environment, this is not a problem, but can result unexpected errors in production environment.
+
+![Chrome Network Throttling](/img/unrealiable-network/chrome_network_throttling2.png)
+
+Chrome and Firefox have good network throttling tools. On Chrome, the throttling tools can be found on the Network tab, while Firefox keeps them in the responsive design mode view. These tools help you to simulate slow network connections or disconnected connection. Unfortunately these tools do not contain a feature of testing randomly failing requests, but at least you can hit the Offline-button to simulate disconnected network during the use of the application.
+
+# TLDR
+
+Reduce the amount of network requests, combine them, re-try failed requests and cache the successful ones. And do not forget to test your application with unreliable network connection.
```

Comment:
Handling failures is a subtle art where many things depend on your app architecture, the details of how the errors manifest (is it simply % of requests failing, or does the duration or size affect failure rate) and the desired user experience.